zerver/tests/test_auth_backends.py

```diff
@@ -1585,6 +1585,87 @@ def test_fetch_auth_backend(self) -> None:
                     'zulip_version': ZULIP_VERSION,
                 })
 
+class TestTwoFactor(ZulipTestCase):
+    def test_two_factor_login(self) -> None:
+        email = self.example_email('hamlet')
+        password = 'testing'
+
+        user_profile = self.example_user('hamlet')
+        user_profile.set_password(password)
+        user_profile.save()
+        self.create_default_device(user_profile)
+
+        with self.settings(TWO_FACTOR_CALL_GATEWAY='two_factor.gateways.fake.Fake',
+                           TWO_FACTOR_SMS_GATEWAY='two_factor.gateways.fake.Fake'):
+            data = {'direct_email': email}
+            result = self.client_post('/accounts/login/local/', data)
+            self.assertEqual(result.status_code, 302)
+            self.assertEqual(get_session_dict_user(self.client.session), user_profile.id)
```

Comment:
What is this actually testing?  The single-HTTP-request flow doesn't seem to match what happens on web...

Reply:
This is testing direct dev. login.

Reply:
Improved it.